### DISCUSSION TITLE
Complete and Look should share semantics/implementation

### DISCUSSION THREAD

**Discussion by paul-lalonde** [rjkroege/edwood#491] [General]
I've been making Look behave more correctly on Windows where path names are much uglier than in unix-land.
Right now, the windows path uses a regexp to match filenames (in expandfile).  That differs from how the posix paths work and how Complete works.  
Probably Complete should handle expandfile as well.
One benefit of how this is done is that expandfile could walk the file tree properly instead of using a stack of heuristics to figure out if something is a filename.  Perhaps heuristics to find the start of a filename, then path following with Complete until a filename is fully identified.
Thoughts?

**Comment by rjkroege** [comment]
Sounds like a good idea to me.

At risk of over-complicating things: I'd like both complete and look to be configurable. If you're engaged in a non-trivial refactoring effort, can you consider letting the logic be provided via some kind of external helper? Then we could provide connect it to lsp?

**Comment by paul-lalonde** [comment]
I was poking at how to hook up such helpers.  Currently the only pattern for external "help" in acme/edwood is the events file.
It's trivial(!?) to implement custom look handler on the L and l events, though I don't know how to sequence such a helper when some other helper (say win) is already parsing the event stream.  Maybe that's not a problem.
Do we extend that model and chain event handling through the whole row of columns of windows?  Then when a helper doesn't handle the event, a global helper could.  The order global helpers are invoked in will be a thing, but that's likely just opening order.
Thoughts?

**Comment by rjkroege** [comment]
> Maybe that's not a problem.

If Chrome is any indication... it's possibly an invitation for race conditions. Our filesystem implementation should be as scalable as the kernel. 🙂

> Do we extend that model and chain event handling through the whole row of columns of windows

I'm not sure what you mean exactly. I think that we should avoid having some kind of event handling chaining. In particular: I'd like the "which helper handles this" should be O(1) and unambiguous. And we should have a well-tested approach for handling "the helper doesn't answer soon enough" -- I worry a little bit about inserting a two-way IPC on the typing critical path.

So maybe: "completion handler" is a per-window attribute?
